### PR TITLE
MAINT/BUG: Remove out-of-band reference count in PyArray_Newshape, fix memory leak

### PR DESCRIPTION
**Pull request by eric-wieser**
Extracted from #11246

### CODE REVIEW

File: numpy/core/src/multiarray/shape.c [hunk 1]
```diff
@@ -246,7 +247,7 @@ PyArray_Newshape(PyArrayObject *self, PyArray_Dims *newdims,
             if (newcopy == NULL) {
                 return NULL;
             }
-            incref = NPY_FALSE;
+            Py_DECREF(self);
```
Comment (mattip) [inline numpy/core/src/multiarray/shape.c]: What happens when `newcopy == NULL`? The Py_DECREF should be before line 247

Comment (mattip) [reply numpy/core/src/multiarray/shape.c]: There are other places where the code exits early with `return NULL`, so probably better to `goto fail` here and in the other ones as well

File: numpy/core/src/multiarray/shape.c [hunk 2]
```diff
@@ -277,9 +278,6 @@ PyArray_Newshape(PyArrayObject *self, PyArray_Dims *newdims,
         goto fail;
     }
 
-    if (incref) {
-        Py_INCREF(self);
-    }
     if (PyArray_SetBaseObject(ret, (PyObject *)self)) {
         Py_DECREF(ret);
         return NULL;
```
Comment (mattip) [inline numpy/core/src/multiarray/shape.c]: missing decref, better as `goto fail`

Comment (eric-wieser) [reply numpy/core/src/multiarray/shape.c]: `PyArray_SetBaseObject` steals a reference, so this is correct as is

File: numpy/core/src/multiarray/shape.c (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/src/multiarray/shape.c]: Same thing as what happened before this patch. Good catch though - I guess this now needs a bug label too.

**Review by mattip — APPROVED** [review]

**Comment by mhvk** [comment]
Looks good to me too; definitely clearer.

**Comment by mattip** [comment]
hmm, since I merged #11248 this needs a rebase

**Comment by mattip** [comment]
Thanks @eric-wieser

**Pull request merged**